mapsims/cmb.py

```diff
@@ -0,0 +1,81 @@
+import healpy as hp, numpy as np
+from pixell import enmap, curvedsky
+
+def get_cmb_sky(iteration_num, 
+                nside = None, #set this if healpix is desired
+                shape = None, #set shape and wcs if CAR maps are desired
+                wcs = None,
+                lensed = True,
+                aberrated = False,
+                pol = True,
+                cmb_set = 0, #We allow for more than one CMB map per lensing map
+                cmb_dir = None, 
+                nfreqs = 1):
+    '''
+    Return a CMB map from stored alm's.  This can be in Healpix format
+    (if nside is specified) or rectangular pixel format (if wcs and shape are
+    specified).  The lensed alm's are pre-stored.
+    If rectangular, it returns a stack of enmaps of shape (nfreqs, ncomp, ny, nx). 
+    If Healpix, it will return a numpy array of shape (nfreqs, ncomp, npix) 
+
+    Args:
+        iteration_num: integer specifying which sim iteration to use
+        nside: nside of healpix map to project alms to. If None, uses 
+        rectangular pixel geometry specified through shape and wcs.
+        shape: shape of ndmap array (see pixell.enmap)
+        wcs: World Coordinate System for geometry of map (see pixell.enmap)
+        lensed: whether to load lensed or unlensed sims
+        aberrated: whether to load aberrated or unaberrated sims
+        pol: if True, return ncomp=3 I,Q,U components, else just ncomp=1 I
+        cmb_set: integer specifying which set of sims to use
+        cmb_dir: override the default lensed alm directory path
+        nfreqs: number of copies of the CMB sky to provide. When modulation
+        is implemented, this argument will be changed to a frequency bandpass
+        specification that applies frequency-dependent modulation to the sims.
+
+    Returns:
+        output: (nfreqs,ncomp,npix) healpix array if nside is specified, else
+        returns (nfreqs,ncomp,Ny,Nx) rectangular pixel ndmap.
+        
+    '''
+    ncomp = 3 if pol else 1
+    filename = _get_cmb_map_string(cmb_dir,iteration_num,cmb_set,lensed,aberrated)
+    #The hdu = (1, 2,3) means get all of T, E, B
+    #Note the alm's are stored as complex32, so upgrade this for processing
+    alm_teb = np.complex128(hp.fitsfunc.read_alm(filename, hdu = (1,2,3) if pol else 1))
+    #Here we can multiply the alms by the appropriate beam;
+    #hp.sphtfunc.almxfl can be used.  Not included yet.  Once we
+    #do, we will have to call the inverse SHT nfreqs times, as in actsims.
+    if nside is not None:
+        #Then we are outputting a healpix map
+        map_tqu = hp.alm2map(alm_teb, nside)
+        output = np.tile(map_tqu, (nfreqs, 1, 1)) # if nfreqs>1 else map_tqu
+        #Here we want to multiply the map by the modulation factor.  FIXME: not implemented yet
```

Comment:
can you explain here what do you mean by a modulation factor?

Reply:
It comes from special relativistic effects from our motion -- each CMB map is multiplied by a position-dependent factor of order (1 + v/c), but the factor depends on frequency.  This means that since the CMB maps depend on frequency (even when in uK_CMB units) we need to design the code to allow for different frequency outputs.

Reply:
So instead of nfreqs, we'll need some way to specify the frequency bandpasses in the interface to this function right? Should we go with some effective central frequency? Or do we want to do a bandpass integration? What will PySM be doing?

Reply:
I am actually now thinking that this model should go into the `so_pysm_models` repo.
It should implement the same interface of the class for synchrotron:
https://github.com/simonsobs/so_pysm_models/pull/1/files#diff-784071308a8ee543573ec589d5a4a094

basically you inizialize it in the contructor and then provide a `signal` method that takes the array of frequencies `nu` and returns the maps at those frequencies.

A similar model I implemented (reads maps from disk and interpolates them): https://github.com/zonca/pysm_interpolating_component/blob/master/__init__.py

Do you think this interface would be suitable?

We can call it `PrecomputedLensedCMB` or something like that

Reply:
I would be OK to move this into PySm or a wrapper thereof.  (In actsims we do something similar for when we need to generate gaussian random fields on the fly.)

Keep in mind though:
a) we would need to retain support for the rectangular pixels, this is important for those working on the LAT survey analysis.  Can PySM handle CAR maps?  I missed the last FG call so I don't know what the plan is there.

b) the modulation factor is a small effect and so the bandpass integration for it does not need to be done to the same precision as the bandpass integration over models of e.g. Galactic dust.

Reply:
a) If we implement the code in a `PrecomputedLensedCMB` class, we are not bind to the PySM, we can just run this class outside of PySM and use whatever pixelization we want.
Also, I think that if we only use components that are all in CAR, PySM should mostly work fine, and if we hit some issues, I can work on fixing them.

b) sorry, I need to see the modulation factor implemented to have a clearer understanding of this.

So can you go ahead and make a pull request to the repo https://github.com/simonsobs/so_pysm_models/ ?

Reply:
Done.